### PR TITLE
Nael/pending rpc requests

### PR DESCRIPTION
**Pull request by nael-binary**

### CODE REVIEW

File: lib/Myriad/RPC/Implementation/Redis.pm [hunk 1]
```diff
@@ -84,16 +86,43 @@ async method stop () {
 async method create_group ($rpc) {
     unless ($rpc->{group}) {
         await $self->redis->create_group($rpc->{stream}, $self->group_name);
+        await $self->process_pending($rpc);
```
Comment (eyadof) [inline lib/Myriad/RPC/Implementation/Redis.pm:89]: Maybe for the sake of clarity we change the sub name into something else?

File: lib/Myriad/Service/Attributes.pm [hunk 2]
```diff
@@ -87,6 +87,8 @@ This will cause the method to be registered in L<Myriad::Registry/add_rpc>.
 
 sub rpc {
     my ($class, $pkg, $method, $code, $args) = @_;
+    # Set default $args if not passed
+    $args = { process_pending => 0 } unless $args and exists $args->{process_pending};
```
Comment (eyadof) [inline lib/Myriad/Service/Attributes.pm:91]: This is a bit of nitpicking, but `process_pending` is a bit of Redis terminology, maybe another one-word term?

File: lib/Myriad/RPC/Implementation/Redis.pm [hunk 3]
```diff
@@ -84,16 +86,43 @@ async method stop () {
...
         $rpc->{group} = 1;
     }
 }
 
+async method process_pending ($rpc) {
+    my $pending_info = await $self->redis->pending_messages_info($rpc->{stream}, $self->group_name);
+    while ($pending_info->[0] > 0) {
```
Comment (eyadof) [inline lib/Myriad/RPC/Implementation/Redis.pm:96]: I think a big backlog of pending messages is a concern here, what should we do? 
- mix pending and new messages randomly? 
- process pending messages when there are no new messages.
- add an option for the developer to drop processing if the backlog is more than a certain threshold?

File: lib/Myriad/RPC/Implementation/Redis.pm [hunk 4]
```diff
@@ -84,16 +86,43 @@ async method stop () {
...
+        my @items = await $self->redis->pending(stream => $rpc->{stream}, group => $self->group_name, client => $self->whoami);
+
+        for my $item (@items) {
+            push $item->{data}->@*, ('transport_id', $item->{id});
+            if ($rpc->{process_pending}){
+                try {
+                    $log->infof('processing pending request %s',$item);
+                    my $message = Myriad::RPC::Message::from_hash($item->{data}->@*);
+                    $rpc->{sink}->emit($message);
```
Comment (eyadof) [inline lib/Myriad/RPC/Implementation/Redis.pm:105]: I think we should start checking `$message->deadline` here (also in the regular delivery)

File: lib/Myriad/RPC/Implementation/Redis.pm (hunk 3 above)
Comment (tom-binary) [reply lib/Myriad/RPC/Implementation/Redis.pm:96]: If there's a big backlog of pending messages, then we process those first.

**Pull request merged**